### ISSUE TITLE
DDP AMP flag is not broadcast correctly when check_amp() disables AMP on rank 0

### ISSUE THREAD

**Issue by LZ-QWQ** [ultralytics/ultralytics#24798]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Multi-GPU

### Bug

When training with DDP, `check_amp()` is [only executed on rank 0](https://github.com/ultralytics/ultralytics/blob/8988f1c5d6a21986a3fdb05b53c64173ebdea913/ultralytics/engine/trainer.py#L331-L339): 

If `check_amp()` returns `False` on rank 0, rank 0 correctly disables AMP. However, the current broadcast call does not update `self.amp` on worker ranks:
https://github.com/ultralytics/ultralytics/blob/8988f1c5d6a21986a3fdb05b53c64173ebdea913/ultralytics/engine/trainer.py#L338

`self.amp.int()` creates a temporary tensor. `dist.broadcast()` updates that temporary tensor, but `self.amp` itself remains unchanged on non-zero ranks. As a result, after setup:
```
rank 0: amp=False
rank 1: amp=True
```
This means DDP can run with mixed AMP state across ranks.

### Environment

```
Ultralytics 8.4.60 🚀 Python-3.12.3 torch-2.10.0+rocm7.2.4.git3d3aa833 CUDA:0 (AMD Radeon Graphics, 30576MiB)
Setup complete ✅ (192 CPUs, 503.0 GB RAM, 2707.5/3666.4 GB disk)

OS                     Linux-6.17.0-35-generic-x86_64-with-glibc2.39
Environment            Docker
Python                 3.12.3
Install                pip
Path                   /opt/venv/lib/python3.12/site-packages/ultralytics
RAM                    503.00 GB
Disk                   2707.5/3666.4 GB
CPU                    AMD Ryzen Threadripper PRO 9995WX 96-Cores
CPU count              192
GPU                    AMD Radeon Graphics, 30576MiB
GPU count              4
CUDA                   None
```

#### Minimal reproducer

The script below forces `check_amp()` to return False at runtime. This is only to make the `check_amp()` failure branch reproducible on any 2-GPU machine; it does not modify Ultralytics source files and is not part of the proposed fix.

**Minimal reproducer: `repro_ddp_check_amp_broadcast_bug.py`**
```python
"""Minimal reproducer for Ultralytics DDP AMP flag broadcast.

Run:
    torchrun --standalone --nproc_per_node=2 repro_ddp_check_amp_broadcast_bug.py

Buggy output:
    trainer.amp per rank: [False, True]

Fixed output:
    trainer.amp per rank: [False, False]
"""

import os

import torch
import torch.distributed as dist
from ultralytics import YOLO
from ultralytics.engine import trainer as trainer_module
from ultralytics.utils import checks as checks_module


def fake_check_amp(_model):
    return False


# Reproducer-only runtime patch: force the check_amp-failed branch so this
# issue can be reproduced on machines where the real AMP check passes.
checks_module.check_amp = fake_check_amp
trainer_module.check_amp = fake_check_amp


def show_amp_and_exit(trainer):
    amp = torch.tensor(int(bool(trainer.amp)), device=trainer.device)
    gathered = [torch.zeros_like(amp) for _ in range(dist.get_world_size())]
    dist.all_gather(gathered, amp)

    if dist.get_rank() == 0:
        print("Simulated check_amp() result: False", flush=True)
        print(f"trainer.amp per rank: {[bool(x.item()) for x in gathered]}", flush=True)

    dist.destroy_process_group()
    raise SystemExit(0)


if int(os.environ.get("WORLD_SIZE", "1")) < 2:
    raise SystemExit("Launch with: torchrun --standalone --nproc_per_node=2 repro_ddp_check_amp_broadcast_bug.py")

model = YOLO("yolo26n.yaml")
model.add_callback("on_train_start", show_amp_and_exit)
model.train(
    data="coco8.yaml",
    device=list(range(int(os.environ["WORLD_SIZE"]))),
    imgsz=640,
    amp=True,
    plots=False,
    project="/tmp/ultralytics_ddp_amp_broadcast_repro",
)
```

**Run**
```bash
torchrun --standalone --nproc_per_node=2 repro_ddp_check_amp_broadcast_bug.py
```
**Current output**
```
Simulated check_amp() result: False
trainer.amp per rank: [False, True]
```
**Expected output**
```
Simulated check_amp() result: False
trainer.amp per rank: [False, False]
```

#### Proposed fix
Keep using integer broadcast for compatibility with older PyTorch/Gloo bool broadcast behavior, but broadcast a named tensor and assign the result back to `self.amp`:
```
diff --git a/ultralytics/engine/trainer.py b/ultralytics/engine/trainer.py
--- a/ultralytics/engine/trainer.py
+++ b/ultralytics/engine/trainer.py
@@ -335,7 +335,9 @@ class BaseTrainer:
             self.amp = torch.tensor(check_amp(self.model), device=self.device)
             callbacks.default_callbacks = callbacks_backup  # restore callbacks
         if RANK > -1 and self.world_size > 1:  # DDP
-            dist.broadcast(self.amp.int(), src=0)  # broadcast from rank 0 to all other ranks; gloo errors with boolean
+            amp = self.amp.int()  # broadcast from rank 0 to all other ranks; gloo errors with boolean
+            dist.broadcast(amp, src=0)
+            self.amp = amp.bool()
         self.amp = bool(self.amp)  # as boolean
         self.scaler = (
             torch.amp.GradScaler("cuda", enabled=self.amp) if TORCH_2_4 else torch.cuda.amp.GradScaler(enabled=self.amp)
```

After this change, the minimal reproducer prints:
```
Simulated check_amp() result: False
trainer.amp per rank: [False, False]
```

#### How we found this
We found this bug on AMD Radeon R9700 (gfx1201) GPUs where [fp16 has precision issues](https://github.com/ROCm/rocm-libraries/issues/4998) and Ultralytics `check_amp()` fails. Rank 0 disabled AMP, but other DDP ranks continued using AMP. This caused NaN losses during training.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @LZ-QWQ, thank you for the detailed bug report and for including a clear minimum reproducible example for this DDP AMP broadcast issue in Multi-GPU training 🚀 This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.

Thanks again for already providing a reproducible script, environment details, expected vs current behavior, and a proposed fix 💡 That makes investigation much easier.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

Thanks for debugging this. Can you open a PR with your suggested fix?

**Comment by LZ-QWQ** [comment]
> Hello.
>
> Thanks for debugging this. Can you open a PR with your suggested fix?

Done: https://github.com/ultralytics/ultralytics/pull/24810

**Comment by glenn-jocher** [comment]
Thanks for opening the PR — we’ll continue the review and discussion there.
